### DISCUSSION TITLE
Question: When are you going to add the Webui?

### DISCUSSION THREAD

**Discussion by YotomiY** [MatteoFasulo/Whisper-TikTok#13] [Ideas]
Hi there,
I just wanted to know when you are going to release the webui version of this project.
I also wanted to ask if it would be possible to add a function: The function would be video focused: Music
I thought it would be nice if background music could be added to the video.

`EDITED*
`I would also like to know when the automation for tiktok (self publishing) will be released.

Thank you.

**Comment by MatteoFasulo** [comment]
> Hi there, I just wanted to know when you are going to release the webui version of this project. I also wanted to ask if it would be possible to add a function: The function would be video focused: Music I thought it would be nice if background music could be added to the video.
> 
> `EDITED* `I would also like to know when the automation for tiktok (self publishing) will be released.
> 
> Thank you.

Hi @YotomiY,

Thank you for reaching out and sharing your ideas for the Whisper-TikTok project! I truly appreciate your interest and enthusiasm.

While I am unable to work on a web UI version or the specific video-focused function of adding background music at the moment, I would be more than thrilled to welcome contributions from the community. If you or any other user is interested in developing such functionalities, please feel free to explore and contribute to the project. Open source projects thrive on collaborative efforts, and your input could greatly enhance the capabilities of Whisper-TikTok.

Regarding the automation for TikTok self-publishing, I want to assure you that it's on our radar. The upcoming release, version 2.0, is indeed planning to include integration with the TikTok API for video uploads. This enhancement will provide users with a more streamlined process for sharing their creations directly on the TikTok platform.

Your suggestions are valuable, and I encourage you to keep an eye on our project's updates. If you or anyone from the community decides to take on the challenge of implementing these features, it would be a fantastic contribution to the Whisper-TikTok community.

Thank you for your understanding and for being part of our exciting journey!

Best regards,

~ Matteo

**Comment by YotomiY** [comment]
Hello,
I'm planning to integrate the webui part using Gardio (used in stable diffusion webui) but I don't think I could do the music part. I haven't investigated that domains.
Thanks

**Reply by MatteoFasulo** [reply]
Cool! 🚀

Thanks for planning to integrate the web UI with Gardio—sounds interesting! No worries about the music part.

By the way, I'm currently working on the TikTok upload feature in the "dev" branch, and there's a beta available. Feel free to check it out and share your thoughts! You will just need to clone the repository from that specific branch

Appreciate your contributions to Whisper-TikTok.

**Reply by YotomiY** [reply]
Hello, 
I am super exited for the feature but i had this error trying it 
```
Error uploading to TikTok: 'NoneType' object has no attribute 'split'
```
I don't know if for you is working, but not for me. 
Thank you

**Reply by MatteoFasulo** [reply]
Hi, is the error happening while generating the first video or the second one?

**Reply by YotomiY** [reply]
Both
```
| OK | Uploading to TikTok...                                                                     main.py:233
| OK | Cookie file found                                                                          main.py:560
⠧ Creating video... | [] Authenticating browser with cookies
[::] Create a chrome browser instance in headless mode
[:] | ERROR | Error uploading to TikTok: 'NoneType' object has no attribute 'split'                   main.py:570
| OK | Uploaded to TikTok successfully!
```

**Reply by MatteoFasulo** [reply]
I will fix that asap

**Reply by YotomiY** [reply]
Cool 👍 
Looking forward to the commit!!
Thanks!

**Reply by MatteoFasulo** [reply]
Hi, I made some changes to the code on dev branch, can you tell me if now it is working with these changes?

**Comment by YotomiY** [comment]
Hello,
I'm having trouble making the webui, the problem is that all the project is on one file (main.py). I don't think it's possible to add a webui using gardio but I'm looking for an alternative any ideas?
Thanks

**Reply by MatteoFasulo** [reply]
Hi,

Just to clarify and make sure I understand correctly, is the primary issue that the entire project is contained within a single main.py file, and you're considering whether dividing it into several files with a main class could be a potential solution?

If that's the case, breaking down the code into submodules and utilizing classes for different components can indeed improve the project's organization and structure. It's a common practice to enhance code readability and maintainability. If you have any specific concerns or questions about this approach, feel free to share more details, and we can discuss potential solutions together.

Looking forward to assisting you further!

**Reply by YotomiY** [reply]
Thank you for the reply. I'm going to try to do it. When do you think you will solve the TikTok problem?

**Reply by MatteoFasulo** [reply]
> Thank you for the reply. I'm going to try to do it. When do you think you will solve the TikTok problem?

Already made a commit about that just a couple of minutes ago so try and tell me if it is working now

**Reply by YotomiY** [reply]
Error!
 | ERROR | Error uploading to TikTok: 'NoneType' object has no attribute 'split'                   main.py:582
 | ERROR | Error uploading to TikTok 

And do you now why my output has now underline subtitles:
![image](https://github.com/MatteoFasulo/Whisper-TikTok/assets/115656415/9462c421-939a-47fc-b9f5-f93eb4783f0d)

**Reply by MatteoFasulo** [reply]
Yes, the default argument is using underline sub_format to 'u' which is underline. I'm working on the 'split' issue

**Reply by YotomiY** [reply]
How do I turn it off ? 
Thanks for solving the issue

**Reply by MatteoFasulo** [reply]
Just use bold instead of underline ('b')

**Reply by YotomiY** [reply]
Cool thanks, how is it going with the issue?

**Reply by MatteoFasulo** [reply]
I changed the default sub-format to bold and also modified the file path for the tiktok upload; try to use the new commit on dev

**Reply by YotomiY** [reply]
im going to check it right now and tell you 
thx

**Reply by YotomiY** [reply]
I still have the same error in the new commit. Is it working for you?
```
⠹ Creating video... | [20:14:36] Authenticating browser with cookies
[20:14:36] Create a chrome browser instance in headless mode
[20:14:39] | ERROR | Error uploading to TikTok: 'NoneType' object has no attribute 'split'
```

**Reply by MatteoFasulo** [reply]
Yes, my upload works without errors.

Btw from the source code of `tiktok-uploader` there are only few `.split()` calls:

https://github.com/wkaisertexas/tiktok-uploader/blob/main/src/tiktok_uploader/upload.py

Try to investigate which of them is generating your issue (e.g. just try to change description manually or path and try to do some debugging on it). Unfortunately I cannot reproduce the issue you have so I do not have any idea on where the issue is coming from on that code :/

You can also insert some prints in the code of  `tiktok-uploader` on the folder of your python environment and investigate what is producing the issue

**Reply by YotomiY** [reply]
This is my full error
Do you have any idea why this is happening
```
2023-12-26` 20:14:39,789 - __main__ - ERROR - 'NoneType' object has no attribute 'split'
Traceback (most recent call last):
  File "Whisper-TikTok\main.py", line 578, in upload_tiktok
    upload_video(file, description=description, cookies='cookies.txt',
  File "C:\Users\\anaconda3\Lib\site-packages\tiktok_uploader\upload.py", line 54, in upload_video
    return upload_videos(
           ^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\tiktok_uploader\upload.py", line 103, in upload_videos
    driver = get_browser(name=browser, headless=headless, proxy=proxy, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\tiktok_uploader\browsers.py", line 34, in get_browser
    service = get_service(name=name)
              ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\tiktok_uploader\browsers.py", line 62, in get_service
    return services[name]()
           ^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\tiktok_uploader\browsers.py", line 196, in <lambda>
    'chrome': lambda : ChromeService(ChromeDriverManager().install()),
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\webdriver_manager\chrome.py", line 40, in install
    driver_path = self._get_driver_binary_path(self.driver)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\webdriver_manager\core\manager.py", line 40, in _get_driver_binary_path
    file = self._download_manager.download_file(driver.get_driver_download_url(os_type))
                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\webdriver_manager\drivers\chrome.py", line 32, in get_driver_download_url
    driver_version_to_download = self.get_driver_version_to_download()
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\webdriver_manager\core\driver.py", line 48, in get_driver_version_to_download
    return self.get_latest_release_version()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\\anaconda3\Lib\site-packages\webdriver_manager\drivers\chrome.py", line 64, in get_latest_release_version
    determined_browser_version = ".".join(determined_browser_version.split(".")[:3])
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Reply by MatteoFasulo** [reply]
Seems like an issue with the browser type. Are you on Windows / MacOS or Linux? 

From the examples, the supported browser are:
- 'chrome'
- 'safari',
- 'chromium',
- 'edge',
- 'firefox'

![a](https://github.com/MatteoFasulo/Whisper-TikTok/assets/74818541/c3da433c-9cee-486d-8289-94461a2035d0)

change applied at line 19*

**Reply by YotomiY** [reply]
windows

**Reply by YotomiY** [reply]
where (what file)

**Reply by MatteoFasulo** [reply]
Well then, do you have any of those browsers installed?

- the file is on `main.py`

**Reply by YotomiY** [reply]
yes opera but i can install chrome

**Reply by MatteoFasulo** [reply]
well then opera is the issue, they are not supporting Opera driver (maybe I can do a PR to do that) so please try to install Chrome and see if it is working. It should detect that you are using chrome and use the associated driver

**Reply by YotomiY** [reply]
Its working perfectly
Thanks

**Reply by MatteoFasulo** [reply]
Btw, issue submitted at https://github.com/wkaisertexas/tiktok-uploader/issues/86#issue-2056666941

**Reply by YotomiY** [reply]
cool!

**Reply by YotomiY** [reply]
Could this library be used to program the videos and not publish them directly??
I awser my self : yes!

**Reply by MatteoFasulo** [reply]
Yes, it has also a scheduling feature but I did not get if it was only for TikTok Business account or just for everyone

**Comment by YotomiY** [comment]
Hello!
I have a question, why is the text not centred and how to solve it and how to change the number of words by frame and size
Thanks

**Reply by MatteoFasulo** [reply]
Hi there!

The positioning of the text in Whisper-TikTok is intentional to ensure that it remains visible and legible, especially considering the TikTok logo presence during scrolling. By left-center aligning the text, we aim to prevent it from being covered by the logo and maintain a better user experience.

This image refers to the alignments available in FFMPEG for subtitles

![image](https://github.com/MatteoFasulo/Whisper-TikTok/assets/74818541/f4592605-d1fd-49af-b8a6-fab6533bc99f)


You can modify settings about subtitles like font, size and so on at line 377 in main.py

---

To adjust the number of words per frame and text size, you can make modifications on `word_options` dict (line 423 - main.py). 

```python
word_options = {
        "highlight_words": True,  # keep this True otherwise the custom style won't be applied
        "max_line_count": 1,
        "max_line_width": 40,
        "preserve_segments": False,
    }
```

Feel free to share your settings which looks good. Also tell me if for simplicity it would be better to have such options as CLI arguments instead of making changes to the dict :)

**Reply by YotomiY** [reply]
Thank for your reply!

- I have an idea that I have seen in several places. In my case, I am creating Reddit stories, seen that they are being quite successful in those moments. And one feature that I see that everyone has is that, at the beginning of the video, a PNG appears as if it were a Reddit capture. Do you think you could make the script take the PNG fill it with text that has been put in the JSON, put it at the beginning of the video and hide the subtitles at the beginning

- I also think it would be good if for example in the middle of the video you could add a subscribe video to encourage people to subscribe.

Thx

**Comment by MatteoFasulo** [comment]
> Thank for your reply!
> 
> * I have an idea that I have seen in several places. In my case, I am creating Reddit stories, seen that they are being quite successful in those moments. And one feature that I see that everyone has is that, at the beginning of the video, a PNG appears as if it were a Reddit capture. Do you think you could make the script take the PNG fill it with text that has been put in the JSON, put it at the beginning of the video and hide the subtitles at the beginning
> * I also think it would be good if for example in the middle of the video you could add a subscribe video to encourage people to subscribe.
> 
> Thx

Hi,

In terms of the Web-UI version, there's a [fork where a Web-UI is currently in development](https://github.com/Easypress-ai/Whisper-TikTok2). I'll be keeping an eye on that periodically to stay updated on its progress. It's expected to be a Streamlit web app, which is pretty cool because Streamlit apps can be hosted for free on their public cloud with minimal adjustments. I'll also make an effort to reach out to the developer and engage in a conversation :)

**Comment by YotomiY** [comment]
I have just seen the new commit ! 
It's just perfect for gradio im going to try making the UI asp 
Thanks!

**Reply by MatteoFasulo** [reply]
Hi, 
follow issue #17 for WebUI
